raster-foundry/raster-foundry
# Add user bulk create endpoint

## Overview

This PR:
- adds a nullable `parent` column to `campaigns` table and refer to the `id` field of the same table, so that cloned campaigns know who they are cloned from
- adds a `/api/users/bulk-create` endpoint to create `count` number of users following a name creation type
- these users are created in auth0 as a result, are in DB, have user group roles assigned, and have  their own campaigns/annotation projects/tasks cloned
- the endpoint returns a list of user names

### Checklist

- ~Swagger specification updated~
- ~Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~

## Testing Instructions

- pull down the new `.env` file from development data config bucket, a new auth0 db connection is in there
- `./scripts/load_development_data --download` to get the sample campaign, which has a sample project connected to it. (Feel free to skip it, but remember to: run migrations, create a campaign, and connect some annotation projects to the campaign)
- assemble api server jar and spin it up
- grab a token from frontend, make sure the account you use is admin of the `Public` organization
- run the following to create 5 user: note that the IDs shouldn't need changes if dev DB is updated with the new pgdump; `peudoUserNameType` can be other values too like `GAME_OF_THRONES`, `HOBBIT`, `LORD_OF_THE_RINGS`, `POKEMON`, `RICK_AND_MORTY`, `SUPER_HERO`
```bash
curl --location --request POST 'http://localhost:9091/api/users/bulk-create' \
--header 'Authorization: Bearer <your token goes here>' \
--header 'Content-Type: application/json' \
--data-raw '{
	"count": 5,
	"peudoUserNameType": "HARRY_POTTER",
	"platformId": "31277626-968b-4e40-840b-559d9c67863c",
	"organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
	"campaignId": "5c725cdf-e8af-4001-a3c4-98211d02f43e"
}'
```
- check in auth0 dev tenant to make sure the users are created and are under the correct connection db. make sure platform id and organization id are in user `app_metadata`
- check in your local db that 5 new users are created, that campaigns and annotation projects are cloned, and that the campaign clones point to their parent
